File: 22_专题/前缀与差分/前缀和/二维前缀和/Presum2DMonoid.go
package main

const INF int = 1e18

type Presum2DMonoid[T any] struct {
	preSum   [][]T
	row, col int32
	e        func() T
}

func NewPresum2DMonoid[T any](matrix [][]T, e func() T, op func(T, T) T) *Presum2DMonoid[T] {
	row, col := int32(len(matrix)), int32(len(matrix[0]))
	preSum := make([][]T, row+1)
	for i := range preSum {
		row := make([]T, col+1)
		for j := range row {
			row[j] = e()
		}
		preSum[i] = row
	}
	for r := int32(0); r < row; r++ {
		tmpSum0, tmpSum1 := preSum[r], preSum[r+1]
		tmpM := matrix[r]
		for c := int32(0); c < col; c++ {
			tmpSum1[c+1] = op(op(tmpSum0[c+1], tmpSum1[c]), tmpM[c])
		}
	}
	return &Presum2DMonoid[T]{preSum: preSum, row: row, col: col, e: e}
}

// 查询[0:x, 0:y]的聚合值.
func (p *Presum2DMonoid[T]) Query(x, y int32) T {
	if x <= 0 || y <= 0 {
		return p.e()
	}
	if x > p.row {
		x = p.row
	}
	if y > p.col {
		y = p.col
	}
	return p.preSum[x][y]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// 3148. 矩阵中的最大得分
// https://leetcode.cn/problems/maximum-difference-score-in-a-grid/
func maxScore(grid [][]int) int {
	S := NewPresum2DMonoid(grid, func() int { return INF }, min)
	res := -INF
	row, col := int32(len(grid)), int32(len(grid[0]))
	for i := int32(0); i < row; i++ {
		for j := int32(0); j < col; j++ {
			res = max(res, grid[i][j]-min(S.Query(i, j+1), S.Query(i+1, j)))
		}
	}
	return res
}

File: 22_专题/前缀与差分/前缀和/二维前缀和/Presum2DMonoid.py
from typing import Callable, Generic, List, TypeVar

T = TypeVar("T")


class Presum2DMonoid(Generic[T]):
    __slot__ = "_presum", "_row", "_col", "_e"

    def __init__(self, matrix: List[List[T]], e: Callable[[], T], op: Callable[[T, T], T]):
        row, col = len(matrix), len(matrix[0])
        presum = [[e() for _ in range(col + 1)] for _ in range(row + 1)]
        for r in range(row):
            tmpsum0, tmpsum1 = presum[r], presum[r + 1]
            tmpm = matrix[r]
            for c in range(col):
                tmpsum1[c + 1] = op(op(tmpsum0[c + 1], tmpsum1[c]), tmpm[c])
        self._presum = presum
        self._row = row
        self._col = col
        self._e = e

    def query(self, x: int, y: int) -> T:
        """查询[0:x, 0:y]的聚合值."""
        if x <= 0 or y <= 0:
            return self._e()
        if x > self._row:
            x = self._row
        if y > self._col:
            y = self._col
        return self._presum[x][y]


def min2(x: int, y: int) -> int:
    return x if x < y else y


def max2(x: int, y: int) -> int:
    return x if x > y else y


if __name__ == "__main__":
    # 3148. 矩阵中的最大得分
    # https://leetcode.cn/problems/maximum-difference-score-in-a-grid/description/
    # 给你一个由 正整数 组成、大小为 m x n 的矩阵 grid。
    # 你可以从矩阵中的任一单元格移动到另一个位于正下方或正右侧的任意单元格（不必相邻）。
    # 从值为 c1 的单元格移动到值为 c2 的单元格的得分为 c2 - c1 。
    # 你可以从 任一 单元格开始，并且必须至少移动一次。
    # 返回你能得到的 最大 总得分。
    class Solution:
        def maxScore(self, grid: List[List[int]]) -> int:
            INF = int(1e18)
            S = Presum2DMonoid(grid, lambda: INF, min2)
            res = -INF
            for i in range(len(grid)):
                for j in range(len(grid[0])):
                    res = max(res, grid[i][j] - min2(S.query(i, j + 1), S.query(i + 1, j)))
            return res
